File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys

input = sys.stdin.readline

K, N = map(int, input().split())
cable = []
for _ in range(K):
    cable.append(int(input()))
start = 1
end = (2 ** 31) - 1
temp = 0
while True:
    mid = (start+end)//2
    if mid == temp:
        break
    temp = mid
    cnt = 0
    for i in cable:
        cnt += i//mid
    if cnt < N:
        end = mid-1
    elif cnt >= N:
        start = mid+1

print(mid)

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
    static int K, N;
    static long temp,start, end,mid;
    static ArrayList<Integer> cable = new ArrayList<>();
    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        K = Integer.parseInt(st.nextToken());
        N = Integer.parseInt(st.nextToken());
        for (int i = 0; i<K; i++){
            cable.add(Integer.parseInt(br.readLine()));
        }
        int cnt;
        start = 1;
        end = Integer.MAX_VALUE;
        temp = 0;
        while(true){
            cnt = 0;
            mid = (long)(start + end)/2;
            if(temp == mid) break;
            temp = mid;
            for(int c : cable){
                cnt += (c/mid);
            }
            if (cnt < N) end = mid-1;

            else if(cnt >= N){ start = mid+1;}
        }
        System.out.println(mid);

    }

}
